// electron-builder/out/electron-builder/src/util/asarUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.checkFileInArchive = exports.AsarPackager = exports.ELECTRON_COMPILE_SHIM_FILENAME = undefined;

var _bluebirdLst;

function _load_bluebirdLst() {
    return _bluebirdLst = require("bluebird-lst");
}

var _bluebirdLst2;

function _load_bluebirdLst2() {
    return _bluebirdLst2 = _interopRequireDefault(require("bluebird-lst"));
}

/** @internal */
let checkFileInArchive = exports.checkFileInArchive = (() => {
    var _ref2 = (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* (asarFile, relativeFile, messagePrefix) {
        function error(text) {
            return new Error(`${messagePrefix} "${relativeFile}" in the "${asarFile}" ${text}`);
        }
        let fs;
        try {
            fs = yield (0, (_asar || _load_asar()).readAsar)(asarFile);
        } catch (e) {
            throw error(`is corrupted: ${e}`);
        }
        let stat;
        try {
            stat = fs.getFile(relativeFile);
        } catch (e) {
            const fileStat = yield (0, (_fs || _load_fs()).statOrNull)(asarFile);
            if (fileStat == null) {
                throw error(`does not exist. Seems like a wrong configuration.`);
            }
            // asar throws error on access to undefined object (info.link)
            stat = null;
        }
        if (stat == null) {
            throw error(`does not exist. Seems like a wrong configuration.`);
        }
        if (stat.size === 0) {
            throw error(`is corrupted: size 0`);
        }
    });

    return function checkFileInArchive(_x2, _x3, _x4) {
        return _ref2.apply(this, arguments);
    };
})();


exports.copyFileOrData = copyFileOrData;

var _electronBuilderUtil;

function _load_electronBuilderUtil() {
    return _electronBuilderUtil = require("electron-builder-util");
}

var _fs;

function _load_fs() {
    return _fs = require("electron-builder-util/out/fs");
}

var _fsExtraP;

function _load_fsExtraP() {
    return _fsExtraP = require("fs-extra-p");
}

var _path = _interopRequireWildcard(require("path"));

var _asar;

function _load_asar() {
    return _asar = require("../asar");
}

var _fileTransformer;

function _load_fileTransformer() {
    return _fileTransformer = require("../fileTransformer");
}

var _asyncTaskManager;

function _load_asyncTaskManager() {
    return _asyncTaskManager = require("./asyncTaskManager");
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const isBinaryFile = (_bluebirdLst2 || _load_bluebirdLst2()).default.promisify(require("isbinaryfile"));
const pickle = require("chromium-pickle-js");
const NODE_MODULES_PATTERN = `${_path.sep}node_modules${_path.sep}`;
const BOWER_COMPONENTS_PATTERN = `${_path.sep}bower_components${_path.sep}`;
/** @internal */
const ELECTRON_COMPILE_SHIM_FILENAME = exports.ELECTRON_COMPILE_SHIM_FILENAME = "__shim.js";
function addValue(map, key, value) {
    let list = map.get(key);
    if (list == null) {
        list = [value];
        map.set(key, list);
    } else {
        list.push(value);
    }
}
function copyFileOrData(fileCopier, data, src, destination, stats) {
    if (data == null) {
        return fileCopier.copy(src, destination, stats);
    } else {
        return (0, (_fsExtraP || _load_fsExtraP()).writeFile)(destination, data);
    }
}
/** @internal */
class AsarPackager {
    constructor(src, destination, options, unpackPattern) {
        this.src = src;
        this.options = options;
        this.unpackPattern = unpackPattern;
        this.fs = new (_asar || _load_asar()).AsarFilesystem(this.src);
        this.electronCompileCache = null;
        this.outFile = _path.join(destination, "app.asar");
    }
    // sort files to minimize file change (i.e. asar file is not changed dramatically on small change)
    pack(isElectronCompile, packager, fileWalker, fileCopierHelper) {
        var _this = this;

        return (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* () {
            let files = yield fileCopierHelper.collect(fileWalker);
            _this.metadata = fileWalker.metadata;
            _this.transformedFiles = fileCopierHelper.transformedFiles;
            // transform before electron-compile to avoid filtering (cache files in any case should be not transformed)
            if (isElectronCompile) {
                files = yield _this.compileUsingElectronCompile(files, packager);
            }
            yield _this.createPackageFromFiles(_this.options.ordering == null ? files : yield _this.order(files), packager.info);
        })();
    }
    compileUsingElectronCompile(files, packager) {
        var _this2 = this;

        return (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* () {
            (0, (_electronBuilderUtil || _load_electronBuilderUtil()).log)("Compiling using electron-compile");
            const metadata = _this2.metadata;
            _this2.electronCompileCache = yield packager.getTempFile("electron-compile-cache");
            const cacheDir = _path.join(_this2.electronCompileCache, ".cache");
            // clear and create cache dir
            yield (0, (_fsExtraP || _load_fsExtraP()).ensureDir)(cacheDir);
            const compilerHost = yield (0, (_fileTransformer || _load_fileTransformer()).createElectronCompilerHost)(_this2.src, cacheDir);
            const nextSlashIndex = _this2.src.length + 1;
            // pre-compute electron-compile to cache dir - we need to process only subdirectories, not direct files of app dir
            yield (_bluebirdLst2 || _load_bluebirdLst2()).default.map(files, function (file) {
                if (file.indexOf(NODE_MODULES_PATTERN) !== -1 || file.indexOf(BOWER_COMPONENTS_PATTERN) !== -1 || !(file.indexOf(_path.sep, nextSlashIndex) !== -1) // ignore not root files
                || !metadata.get(file).isFile()) {
                    return null;
                }
                return compilerHost.compile(file).then(function (it) {
                    return null;
                });
            }, (_fs || _load_fs()).CONCURRENCY);
            yield compilerHost.saveConfiguration();
            const cacheFiles = yield (0, (_fs || _load_fs()).walk)(cacheDir, function (file, stat) {
                return !file.startsWith(".");
            }, {
                consume: function (file, fileStat) {
                    _this2.metadata.set(file, fileStat);
                    return null;
                }
            });
            // add shim
            const shimPath = `${_this2.src}/${ELECTRON_COMPILE_SHIM_FILENAME}`;
            cacheFiles.push(shimPath);
            metadata.set(shimPath, { isFile: function () {
                    return true;
                }, isDirectory: function () {
                    return false;
                } });
            _this2.transformedFiles = new Array(cacheFiles.length).concat(_this2.transformedFiles);
            _this2.transformedFiles[cacheFiles.length - 1] = `
'use strict';
require('electron-compile').init(__dirname, require('path').resolve(__dirname, '${packager.info.metadata.main || "index"}'), true);
`;
            // cache files should be first (better IO)
            return cacheFiles.concat(files);
        })();
    }
    detectUnpackedDirs(files, autoUnpackDirs, unpackedDest) {
        var _this3 = this;

        return (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* () {
            const dirToCreate = new Map();
            const metadata = _this3.metadata;
            /* tslint:disable:rule1 prefer-const */
            for (let i = 0, n = files.length; i < n; i++) {
                const file = files[i];
                const index = file.lastIndexOf(NODE_MODULES_PATTERN);
                if (index < 0) {
                    continue;
                }
                let nextSlashIndex = file.indexOf(_path.sep, index + NODE_MODULES_PATTERN.length + 1);
                if (nextSlashIndex < 0) {
                    continue;
                }
                if (file[index + NODE_MODULES_PATTERN.length] === "@") {
                    nextSlashIndex = file.indexOf(_path.sep, nextSlashIndex + 1);
                }
                if (!metadata.get(file).isFile()) {
                    continue;
                }
                const packageDir = file.substring(0, nextSlashIndex);
                if (autoUnpackDirs.has(packageDir)) {
                    const fileParent = _path.dirname(file);
                    if (fileParent !== packageDir && !autoUnpackDirs.has(fileParent)) {
                        autoUnpackDirs.add(fileParent);
                        addValue(dirToCreate, _this3.getRelativePath(packageDir), _path.relative(packageDir, fileParent));
                    }
                    continue;
                }
                const ext = _path.extname(file);
                let shouldUnpack = false;
                if (ext === ".dll" || ext === ".exe") {
                    shouldUnpack = true;
                } else if (ext === "") {
                    shouldUnpack = yield isBinaryFile(file);
                }
                if (!shouldUnpack) {
                    continue;
                }
                (0, (_electronBuilderUtil || _load_electronBuilderUtil()).debug)(`${_this3.getRelativePath(packageDir)} is not packed into asar archive - contains executable code`);
                let fileParent = _path.dirname(file);
                // create parent dir to be able to copy file later without directory existence check
                addValue(dirToCreate, _this3.getRelativePath(packageDir), _path.relative(packageDir, fileParent));
                while (fileParent !== packageDir) {
                    autoUnpackDirs.add(fileParent);
                    fileParent = _path.dirname(fileParent);
                }
                autoUnpackDirs.add(packageDir);
            }
            if (dirToCreate.size > 0) {
                // child directories should be not created asynchronously - parent directories should be created first
                yield (_bluebirdLst2 || _load_bluebirdLst2()).default.map(dirToCreate.keys(), (() => {
                    var _ref = (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* (it) {
                        const base = _path.join(unpackedDest, it);
                        yield (0, (_fsExtraP || _load_fsExtraP()).ensureDir)(base);
                        yield (_bluebirdLst2 || _load_bluebirdLst2()).default.each(dirToCreate.get(it), function (it) {
                            return (0, (_fsExtraP || _load_fsExtraP()).ensureDir)(_path.join(base, it));
                        });
                    });

                    return function (_x) {
                        return _ref.apply(this, arguments);
                    };
                })(), (_fs || _load_fs()).CONCURRENCY);
            }
        })();
    }
    getTargetPath(p, to) {
        if (this.electronCompileCache != null && p.startsWith(this.electronCompileCache)) {
            return this.electronCompileCache.replace(this.electronCompileCache, to);
        }
        return p.replace(this.src, to);
    }
    getRelativePath(p) {
        if (this.electronCompileCache != null && p.startsWith(this.electronCompileCache)) {
            return _path.relative(this.electronCompileCache, p);
        }
        return _path.relative(this.src, p);
    }
    createPackageFromFiles(files, packager) {
        var _this4 = this;

        return (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* () {
            const metadata = _this4.metadata;
            // search auto unpacked dir
            const unpackedDirs = new Set();
            const unpackedDest = `${_this4.outFile}.unpacked`;
            yield (0, (_fsExtraP || _load_fsExtraP()).ensureDir)(_path.dirname(_this4.outFile));
            if (_this4.options.smartUnpack !== false) {
                yield _this4.detectUnpackedDirs(files, unpackedDirs, unpackedDest);
            }
            const dirToCreateForUnpackedFiles = new Set(unpackedDirs);
            const transformedFiles = _this4.transformedFiles;
            const taskManager = new (_asyncTaskManager || _load_asyncTaskManager()).AsyncTaskManager(packager.cancellationToken);
            const fileCopier = new (_fs || _load_fs()).FileCopier();
            for (let i = 0, n = files.length; i < n; i++) {
                const file = files[i];
                const stat = metadata.get(file);
                const relativePath = _this4.getRelativePath(file);
                if (stat != null && stat.isFile()) {
                    const fileParent = _path.dirname(file);
                    const dirNode = _this4.fs.getOrCreateNode(_this4.getRelativePath(fileParent));
                    const newData = transformedFiles == null ? null : transformedFiles[i];
                    const node = _this4.fs.getOrCreateNode(relativePath);
                    node.size = newData == null ? stat.size : Buffer.byteLength(newData);
                    if (dirNode.unpacked || _this4.unpackPattern != null && _this4.unpackPattern(file, stat)) {
                        node.unpacked = true;
                        if (newData != null) {
                            transformedFiles[i] = null;
                        }
                        if (!dirNode.unpacked && !dirToCreateForUnpackedFiles.has(fileParent)) {
                            dirToCreateForUnpackedFiles.add(fileParent);
                            yield (0, (_fsExtraP || _load_fsExtraP()).ensureDir)(_this4.getTargetPath(fileParent, unpackedDest));
                        }
                        const unpackedFile = _this4.getTargetPath(file, unpackedDest);
                        taskManager.addTask(copyFileOrData(fileCopier, newData, file, unpackedFile, stat));
                        if (taskManager.tasks.length > (_fs || _load_fs()).MAX_FILE_REQUESTS) {
                            yield taskManager.awaitTasks();
                        }
                    } else {
                        if (newData == null) {
                            transformedFiles[i] = true;
                        }
                        _this4.fs.insertFileNode(node, stat, file);
                    }
                } else if (stat == null || stat.isDirectory()) {
                    let unpacked = false;
                    if (unpackedDirs.has(file)) {
                        unpacked = true;
                    } else {
                        for (const dir of unpackedDirs) {
                            if (file.length > dir.length + 2 && file[dir.length] === _path.sep && file.startsWith(dir)) {
                                unpacked = true;
                                unpackedDirs.add(file);
                                // not all dirs marked as unpacked after first iteration - because node module dir can be marked as unpacked after processing node module dir content
                                // e.g. node-notifier/example/advanced.js processed, but only on process vendor/terminal-notifier.app module will be marked as unpacked
                                yield (0, (_fsExtraP || _load_fsExtraP()).ensureDir)(_this4.getTargetPath(file, unpackedDest));
                                break;
                            }
                        }
                    }
                    _this4.fs.insertDirectory(relativePath, unpacked);
                } else if (stat.isSymbolicLink()) {
                    _this4.fs.getOrCreateNode(relativePath).link = stat.relativeLink;
                }
            }
            if (taskManager.tasks.length > 0) {
                yield taskManager.awaitTasks();
            }
            yield _this4.writeAsarFile(files);
        })();
    }
    writeAsarFile(files) {
        const headerPickle = pickle.createEmpty();
        headerPickle.writeString(JSON.stringify(this.fs.header));
        const headerBuf = headerPickle.toBuffer();
        const sizePickle = pickle.createEmpty();
        sizePickle.writeUInt32(headerBuf.length);
        const sizeBuf = sizePickle.toBuffer();
        const transformedFiles = this.transformedFiles;
        const writeStream = (0, (_fsExtraP || _load_fsExtraP()).createWriteStream)(this.outFile);
        return new (_bluebirdLst2 || _load_bluebirdLst2()).default((resolve, reject) => {
            writeStream.on("error", reject);
            writeStream.on("close", resolve);
            writeStream.write(sizeBuf);
            const w = index => {
                let data;
                while (true) {
                    if (index >= files.length) {
                        writeStream.end();
                        return;
                    }
                    if ((data = transformedFiles[index++]) != null) {
                        break;
                    }
                }
                const file = files[index - 1];
                if (data !== true) {
                    writeStream.write(data, () => w(index));
                    return;
                }
                const readStream = (0, (_fsExtraP || _load_fsExtraP()).createReadStream)(file);
                readStream.on("error", reject);
                readStream.once("end", () => w(index));
                readStream.pipe(writeStream, {
                    end: false
                });
            };
            writeStream.write(headerBuf, () => w(0));
        });
    }
    order(filenames) {
        var _this5 = this;

        return (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* () {
            const orderingFiles = (yield (0, (_fsExtraP || _load_fsExtraP()).readFile)(_this5.options.ordering, "utf8")).split("\n").map(function (line) {
                if (line.indexOf(":") !== -1) {
                    line = line.split(":").pop();
                }
                line = line.trim();
                if (line[0] === "/") {
                    line = line.slice(1);
                }
                return line;
            });
            const ordering = [];
            for (const file of orderingFiles) {
                const pathComponents = file.split(_path.sep);
                for (const pathComponent of pathComponents) {
                    ordering.push(_path.join(_this5.src, pathComponent));
                }
            }
            const sortedFiles = [];
            let missing = 0;
            const total = filenames.length;
            for (const file of ordering) {
                if (!(sortedFiles.indexOf(file) !== -1) && filenames.indexOf(file) !== -1) {
                    sortedFiles.push(file);
                }
            }
            for (const file of filenames) {
                if (!(sortedFiles.indexOf(file) !== -1)) {
                    sortedFiles.push(file);
                    missing += 1;
                }
            }
            (0, (_electronBuilderUtil || _load_electronBuilderUtil()).log)(`Ordering file has ${(total - missing) / total * 100}% coverage.`);
            return sortedFiles;
        })();
    }
}exports.AsarPackager = AsarPackager;
